www/blog/2024-05-23-trpc-actions.mdx

```diff
@@ -0,0 +1,265 @@
+    // Do something with the input
+  });
+
+// Since we're using the `experimental_caller`,
+// our procedure is now just an ordinary function:
+createPost;
+// ^?
+```
+
+Wow, it's that easy to define a server action that's protected from unauthenticated users, with input validation to protect against attacks such as SQL injections. Let's import this function on the client and call it.
+
+```tsx title="app/post-form.tsx"
+'use client';
+
+import { createPost } from '~/_actions';
+
+export function PostForm() {
+  return (
+    <form
+      // Use `action` to make form progressively enhanced
+      action={createPost}
+      // `Using `onSubmit` allows building rich interactive
+      // forms once JavaScript has loaded
+      onSubmit={async (e) => {
+        e.preventDefault();
+        const title = new FormData(e.target).get('title');
+        // Maybe show loading toast, etc etc. Endless possibilities
+        await createPost({ title });
+      }}
+    >
+      <input type="text" name="title" />
+      <button type="submit">Create Post</button>
+    </form>
+  );
+}
+```
+
+## Going further
+
+Using tRPC builders and it's composable way of defining reusable procedures, we can easily build more complex server actions. Below are some examples:
+
+### Rate Limiting
+
+You can use a service such as Unkey to rate limit your server actions:
+
+```ts twoslash title="server/trpc.ts"
+// TODO
+```
+
+### Observability
+
+You can use `@baselime/node-opentelemtry`'s trpc plugin to add observability in just a few lines of code:
+
+```diff
+--- server/trpc.ts
++++ server/trpc.ts
++ import { tracing } from '@baselime/node-opentelemetry/trpc';
+
+  export const serverActionProcedure = t.procedure
+    .experimental_caller(
+      experimental_nextAppDirCaller({
+        pathExtractor: (meta: Meta) => meta.span,
+      }),
+    )
+    .use(async (opts) => {
+      // Inject user into context
+      const user = await currentUser();
+      return opts.next({ ctx: { user } });
+    })
++  .use(tracing());
+
+--- app/_actions.ts
++++ app/_actions.ts
+  export const createPost = protectedAction
++   .meta({ span: 'create-post' })
+    .input(
+      z.object({
+        title: z.string(),
+      }),
+    )
+    .mutation(async ({ ctx, input }) => {
+      // Do something with the input
+    });
+```
+
+The possibilities are endless, and I bet you already got a ton of nice utility middlewares you're using in your tRPC applications today. These new primitives should also be easy to incrementally adopt, so you can move individual procedures over from your existing tRPC API to server actions without having to rewrite your entire API. If you wanna see an app using this in action, check out [Trellix tRPC](https://trellix-trpc.vercel.app), an app I made recently utilizing these new primitives.
+
+So, what do you think? Let us know over at [Github](https://github.com/trpc/trpc/discussions/5737) and help us iterate to get these primitives to a stable state.
```

Comment:
```suggestion
## What do you think? We want your feedback

So, what do you think? Let us know over at [Github](https://github.com/trpc/trpc/discussions/5737) and help us iterate to get these primitives to a stable state.

[...]
```

We can mention some "non solved bits" here

One big area for us to figure out at the moment is how to error handling in a way where it doesn't just throw to the nearest error boundary